datafusion-examples/examples/simple_udtf.rs

```diff
@@ -0,0 +1,224 @@
+    df.show().await?;
+
+    // stream csv table
+    let df2 = ctx
+        .sql(format!("SELECT * FROM read_csv_stream('{csv_file}');").as_str())
+        .await?;
+    df2.show().await?;
+
+    Ok(())
+}
+
+// Option1: (full implmentation of a TableProvider)
+struct LocalCsvTable {
+    schema: SchemaRef,
+    exprs: Vec<Expr>,
+    batches: Vec<RecordBatch>,
+}
+
+#[async_trait]
+impl TableProvider for LocalCsvTable {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn schema(&self) -> SchemaRef {
+        self.schema.clone()
+    }
+
+    fn table_type(&self) -> TableType {
+        TableType::Base
+    }
+
+    async fn scan(
+        &self,
+        state: &SessionState,
+        projection: Option<&Vec<usize>>,
+        _filters: &[Expr],
+        _limit: Option<usize>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        if !self.exprs.is_empty() {
+            self.interpreter_expr(state).await?;
+        }
+        Ok(Arc::new(MemoryExec::try_new(
+            &[self.batches.clone()],
+            TableProvider::schema(self),
+            projection.cloned(),
+        )?))
+    }
+}
+
+impl LocalCsvTable {
+    // TODO(veeupup): maybe we can make interpreter Expr this more simpler for users
+    // TODO(veeupup): maybe we can support more type of exprs
+    async fn interpreter_expr(&self, state: &SessionState) -> Result<()> {
+        use datafusion::logical_expr::expr_rewriter::normalize_col;
+        use datafusion::logical_expr::utils::columnize_expr;
+        let plan = LogicalPlan::EmptyRelation(EmptyRelation {
+            produce_one_row: true,
+            schema: Arc::new(DFSchema::empty()),
+        });
+        let logical_plan = Projection::try_new(
+            vec![columnize_expr(
+                normalize_col(self.exprs[0].clone(), &plan)?,
+                plan.schema(),
+            )],
+            Arc::new(plan),
+        )
+        .map(LogicalPlan::Projection)?;
+        let rbs = collect(
+            state.create_physical_plan(&logical_plan).await?,
+            Arc::new(TaskContext::from(state)),
+        )
+        .await?;
+        println!("time now: {:?}", rbs[0].column(0));
+        Ok(())
+    }
+}
+
+struct LocalCsvTableFunc {}
+
+impl TableFunctionImpl for LocalCsvTableFunc {
+    fn call(&self, exprs: &[Expr]) -> Result<Arc<dyn TableProvider>> {
+        let mut new_exprs = vec![];
+        let mut filepath = String::new();
+        for expr in exprs {
+            match expr {
+                Expr::Literal(ScalarValue::Utf8(Some(ref path))) => {
+                    filepath = path.clone()
+                }
+                expr => new_exprs.push(expr.clone()),
+            }
+        }
+        let (schema, batches) = read_csv_batches(filepath)?;
+        let table = LocalCsvTable {
+            schema,
+            exprs: new_exprs.clone(),
+            batches,
+        };
+        Ok(Arc::new(table))
+    }
+}
+
+// Option2: (use StreamingTable to make it simpler)
```

Comment:
I actually think adding a second option is more confusing in this example, so I would recommend removing this part (perhaps you can just add a note that it is posisble)